### PR TITLE
[core] Push `v_head` when using `AutoModelForCausalLMWithValueHead`

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

This PR addresses saving `v_head` weights when calling `save_pretrained`. Before this PR if a user wanted to share intermediate-trained `v_head` weights it was not possible. The workaround that I have found needs to load the state dict twice, once by the `from_pretrained` method that is called by the `pretrained_model` attribute and once afterwards to manually load the `v_head`.
This PR introduces a new kwarg on the method `from_pretrained` termed as `resume_training` as I think this is quite an edge case.

Added also some tests

TODOs:
- test `push_to_hub`

cc @lvwerra @lewtun

### CODE REVIEW

File: trl/models/modeling_base.py [hunk 1]
```diff
@@ -82,10 +88,31 @@ class and the arguments that are specific to trl models.
...
             )
-
         # Then, create the full model by instantiating the wrapper class
         model = cls(pretrained_model, **trl_model_args)
 
+        # if resume_training, load the state_dict again - this is ok since the
+        # state_dict is removed from the model after loading it.
+        # since this is only done in edge cases we force the user to do it only
+        # if they explicitly set resume_training to True
+        if resume_training:
```
Comment (lvwerra) [inline trl/models/modeling_base.py:98]: If someone is using the `AutoModelForCausalLMWithValueHead` to load a model I think we can assume that they want to use the Value head so I we should always load it if it's available. What do you think?

Comment (younesbelkada) [reply trl/models/modeling_base.py:98]: OK for me! 
We should advice users to use `AutoModelForCausalLM` when using the model outside `trl` (i.e. not for training)

File: docs/source/quickstart.mdx [hunk 2]
```diff
@@ -47,4 +47,23 @@ reward = [torch.tensor(1.0)]
 train_stats = ppo_trainer.step([query_tensor[0]], [response_tensor[0]], reward)
 ```
 
-In general, you would run steps 3-6 in a for-loop and run it on many diverse queries. You can find a more realistic examples in the examples section. 
\ No newline at end of file
+In general, you would run steps 3-6 in a for-loop and run it on many diverse queries. You can find a more realistic examples in the examples section. 
+
+## How to use the model
```
Comment (younesbelkada) [inline docs/source/quickstart.mdx:52]:
```suggestion
## How to use a trained model
```

File: trl/models/modeling_base.py (hunk 1 above)
Comment (younesbelkada) [reply trl/models/modeling_base.py]: Added more description in [853d52d](https://github.com/lvwerra/trl/pull/86/commits/853d52d0e03f891c074a51c081da2a4e75dea40b)

**Review by lvwerra — APPROVED** [review]

**Comment by younesbelkada** [comment]
Actually this was not working in the sharded case, I'd expect users to load large & sharded models too!

**Review by lvwerra — APPROVED** [review]
